servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -126,73 +144,94 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
   @Override
   public BlockWriter getBlockWriter(long userId, long blockId) throws NotFoundException,
       IOException {
-    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-    return new LocalFileBlockWriter(tempBlockMeta);
+    // NOTE: a temp block is supposed to be visible for its own writer, unnecessary to acquire
+    // block lock here since no sharing
+    mMetadataLock.readLock().lock();
+    try {
+      TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+      return new LocalFileBlockWriter(tempBlockMeta);
+    } finally {
+      mMetadataLock.readLock().unlock();
+    }
   }
 
   @Override
   public BlockReader getBlockReader(long userId, long blockId, long lockId)
       throws NotFoundException, InvalidStateException, IOException {
     mLockManager.validateLock(userId, blockId, lockId);
-    BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
-    return new LocalFileBlockReader(blockMeta);
+    mMetadataLock.readLock().lock();
+    try {
+      BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+      return new LocalFileBlockReader(blockMeta);
+    } finally {
+      mMetadataLock.readLock().unlock();
+    }
   }
 
   @Override
   public TempBlockMeta createBlockMeta(long userId, long blockId, BlockStoreLocation location,
       long initialBlockSize) throws AlreadyExistsException, OutOfSpaceException, NotFoundException,
       IOException, InvalidStateException {
-    mEvictionLock.writeLock().lock();
-    try {
-      return createBlockMetaNoLock(userId, blockId, location, initialBlockSize);
-    } finally {
-      mEvictionLock.writeLock().unlock();
+    for (int i = 0; i < MAX_RETRIES; i ++) {
+      TempBlockMeta tempBlockMeta =
+          createBlockMetaInternal(userId, blockId, location, initialBlockSize, true);
+      if (tempBlockMeta != null) {
+        return tempBlockMeta;
+      }
+      // Failed to allocate a temp block, so trigger Evictor to make some space.
+      // NOTE: Successful {@link freeSpaceInternal} here does not ensure the next try of allocation
+      // also successful, because these two operations are not atomic.
+      if (i < MAX_RETRIES - 1) {
+        freeSpaceInternal(userId, initialBlockSize, location);
+      }
```

Comment:
I think you can move freeSpaceInternal before the createBlockMetaInternal since the eviction should no-op if there is already space? Just makes the code easier to reason about I think.